### ISSUE TITLE
`rabbitmq-env.bat` misbehaves if `ERL_LIBS` contains spaces

### ISSUE THREAD

**Issue by ash-lshift** [rabbitmq/rabbitmq-server#859]
We've found that if Rabbit is installed on Windows to a path containing spaces, trying to run the command line tools gives the following error:

```
{"init terminating in do_boot",
 {undef,[{rabbit_nodes,ensure_epmd,[],[]},
         {rabbit_cli,start_distribution,0,
                     [{file,"src/rabbit_cli.erl"},{line,152}]},
         {rabbit_cli,ensure_cli_distribution,0,
                     [{file,"src/rabbit_cli.erl"},{line,49}]},
         {rabbit_cli,main,3,[{file,"src/rabbit_cli.erl"},{line,62}]},
         {init,start_it,1,[{file,"init.erl"},{line,1054}]},
         {init,start_em,1,[{file,"init.erl"},{line,1035}]}]}}
init terminating in do_boot ()
```

Using a path with no spaces but an otherwise identical setup gives no error.

I think this can be attributed to the following snippet in `rabbitmq-env.bat`:

```batch
REM Ensure all paths in ERL_LIBS do not contains non-ASCII characters.
set ERL_LIBS_orig=%ERL_LIBS%
set ERL_LIBS=
call :filter_paths "%ERL_LIBS_orig%"
goto :filter_paths_done

:filter_paths
set paths=%1
set paths=%paths:"=%
rem " [this line added by ash-lshift to fix highlighting...]
for /f "tokens=1* delims=;" %%a in ("%paths%") do (
    if not "%%a" == "" call :filter_path %%a
    if not "%%b" == "" call :filter_paths %%b
)
set paths=
exit /b
:filter_path
set ERL_LIBS=%ERL_LIBS%;%~dps1%~n1%~x1
exit /b
:filter_paths_done
```

Here's a self-contained batch script to demonstrate the problem.

```batch
set ERL_LIBS=
call :filter_paths "c:\program files"
goto :filter_paths_done

:filter_paths
set paths=%1
set paths=%paths:"=%
rem " [this line added by ash-lshift to fix highlighting...]
for /f "tokens=1* delims=;" %%a in ("%paths%") do (
    if not "%%a" == "" call :filter_path %%a
    if not "%%b" == "" call :filter_paths %%b
)
set paths=
exit /b
:filter_path
set ERL_LIBS=%ERL_LIBS%;%~dps1%~n1%~x1
exit /b
:filter_paths_done
echo %ERL_LIBS%
```

This prints `;c:\program` (on my machine at least -- a fresh EC2 instance with Win Server 2012 R2). I think the expected output is `c:\progra~1`, or maybe `c:\program files`. You can also try it with less contrived paths, eg `"c:\program files\rabbitmq server\rabbitmq_server-3.6.1\plugins"` which gives the same output.

By the way I'm not sure what the intention was with the `%~dps1%~n1%~x1`; it converts part of the path to 8.3 format but not all of it. `%~fs1` converts the whole path to 8.3 format.

I think to deal with spaces it's enough to wrap `%%a` and `%%b` in double quotes:

```
if not "%%a" == "" call :filter_path "%%a"
if not "%%b" == "" call :filter_paths "%%b"
```

**Comment by michaelklishin** [comment]
Thank you very much for digging it out.

**Comment by kjnilsson** [comment]
@ash-lshift the only way I can reproduce your error is when I manually run `rabbitmq-env.bat` before `rabbitmqctl.bat`. Does this sound familiar at all? If so you should not need to run `rabbitmq-env.bat` on it's own, it is only to be used by other scripts. I tried it with and without your proposed escaping change with the same result. 

If you open up a fresh instance of the RabbitMQ Command Prompt and only run `rabbitmqctl` does it work then?

**Comment by rhmoult** [comment]
I installed RabbitMQ 3.6.3 in the default location and did the following on a 32 bit Win7 OS:
Start an admin command prompt
cd into the sbin folder under RabbitMQ server installation directory (e.g. C:\Program Files\RabbitMQ Server\rabbitmq_server-3.6.3\sbin)
Run rabbitmq-service.bat remove

I got the following results:  
![rabbitcrash](https://cloud.githubusercontent.com/assets/7614599/16686909/4a24e2f8-44e3-11e6-94b3-c11f4f68133e.png)

**Comment by kjnilsson** [comment]
@rhmoult your error is slightly different. It complains about a syntax error. Have you edited the config file at all?

**Comment by michaelklishin** [comment]
@kjnilsson in theory this could be caused by an incorrect/unexpected set of arguments passed on the command line. Take a look at the topmost directory name, for example.

**Comment by rhmoult** [comment]
Didn't modify anything, but I was able to fix this by uninstalling rabbit and erlang, reinstalling into c:\programdata (no spaces) and then setting ERLANG_HOME to c:\programdata\erl8.0.  Everything seems to be hunky dory now!

**Comment by michaelklishin** [comment]
@rhmoult OK, so there were spaces involved, thanks. On an unrelated note, Erlang 19.0 breaks RabbitMQ management plugin, so please stick to 18.3 for now.

**Comment by michaelklishin** [comment]
I'm afraid we cannot reproduce the original issue without explicitly including `rabbitmq-env.bat` (which is not meant to be included into user scripts).

**Comment by ash-lshift** [comment]
Just tried again with a fresh install of erlang 18.3 & rabbit 3.6.3, same issue. I installed erlang to `D:\dir with spaces\erl` and rabbit to `D:\dir with spaces\rabbit`. (`D:` being an otherwise empty disk.)
Here's what happens when I run `rabbitmqctl status` after removing the `@echo off` lines from the `.bat` files: https://gist.github.com/anonymous/10c22e58f46d7c9997f3c0528a458e98
As you can see, `ERL_LIBS` ends up being set to `;d:\dir`.
I can probably arrange to give you access to the box if that would help.

**Comment by michaelklishin** [comment]
@ash-lshift thank you. So spaces in the Erlang installation directory seem to be key. @kjnilsson have you tried this scenario?

**Comment by kjnilsson** [comment]
Yes my erlang install directory has spaces as does my rabbitmq install directory (all the windows defaults normally have spaces) but this dump gives me something to work with. Will investigate further. Thank you @ash-lshift

**Comment by kjnilsson** [comment]
@ash-lshift could you check the state of 8.3 name creation on your d: drive? `fsutil 8dot3name query d:`

**Comment by kjnilsson** [comment]
Ok I've finally managed to reproduce this. :) I had to disable 8.3 name creation, reboot and remove and reinstall erlang and rabbitmq. 

This replicates the reported error, however it also appears to affect the running of the rabbitmq broker itself. After install the rabbit application isn't running, and `rabbitmq-service.bat start` yields the same error. 

Once I added the extra quotation marks as suggested by @ash-lshift then I was able to start the service and the `rabbitmqctl.bat status` command completed successfully.
